MAGPI PlatformIO/src/main.cpp

```diff
@@ -0,0 +1,461 @@
+/* Info
+#####  Sensors/Input  #####
+Data/Signal -> Sensor <Library> [protocol]
+
+Numerical Data:
+- GPS -> neo GPS <TinyGPS++> [Serial Port8]
+- Accelration, Gyro -> MPU6050 <Adafruit_sensor.h> [I2C]
+- Alt, Pressure, Temp -> BMP390 <Adafurit> [I2C]
+
+Logic Signal:
+- Gear Motor Position/Step -> Encoder <Encoder.h> [Interrupt/ISR]
+
+##### Actuator/Output #####
+Gear Motor with L293D to control line deflection
+L293D I/O
+Channel 1_3A I
+Channel 2_4A I
+Channel EN   I
+Channel 1_3Y O
+Channel 2_4Y O
+
+Logic Table:
+A  EN  Y
+H  H   H
+L  H   L
+X  L   Z (High Impedance)
+
+
+##### Top Level Control Flow (without logic) #####
+get data -> calculate -> run control algorithm -> store data 
+*/
+
+/*Gordon to do
+- Start up: Use serial port???
+- complete logic then add time_out routines [DONE]
+- Algorthim data writing: Encoder, P controller output, every step
+- check sensor sampling rate
+- set sensor range
+
+*/
+
+/* Time Routine
+the data file will look like (subject to change)
+
+GPS time, time since_last_GPS
+### GPS received ###
+10:00:00, 0, data1, data2, .....
+10:00:00, 10, data1, data2, .....
+*/
+#include <Arduino.h>
+#include "Wire.h" //I2C
+#include "Adafruit_Sensor.h" 
+#include "Adafruit_BMP3XX.h"
+#include "Adafruit_MPU6050.h"
+
+#include "SPI.h"
+#include "SD.h"
+
+#define ENCODER_OPTIMIZE_INTERRUPTS
+#include "Encoder.h"
+
+//GPS
+#include <TinyGPS++.h>
+
+//Algorthim
+#include "cmath"
+#include "SimpleKalmanFilter.h"
+
+//states
+bool launched = 0;
+bool deployed = 0;
+bool test_status = 0;
+
+//variables used for Hall effect sensor /////////////need checking///////////
+volatile byte half_revolutions;
+unsigned int rpm;
+unsigned long timeold;
+
+//variables used for GPS
+double lat ,lon, lat_old, lon_old, speed, gps_alt, heading;
+double elasped_time;
+static const double TARGET_LAT = 12.123456, TARGET_LON = 12.123456; ////////////change GPS Const
+unsigned long current_time, gps_time;
+
+double filtered_alt;
+
+//motor
+#define EN 5 //replace pin number
+#define A1_3 3 //L293d channel 1A 3A 
+#define A2_4 4 //
+int motor_position = -999; // to be zero by the the hall effect sensor  
```

Comment:
We are not using the hall sensor any more right? In that case this should start at 0.